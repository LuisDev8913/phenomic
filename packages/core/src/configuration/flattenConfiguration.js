import defaultConfig from "../defaultConfig.js";

import normalizeBaseUrl from "./normalize-base-url.js";

const debug = require("debug")("phenomic:core:configuration");

const normalizePlugin = (plugin: PhenomicPlugin) => {
  if (!plugin) {
    throw new Error("phenomic: You provided an undefined plugin");
  }

  debug("plugin typeof", typeof plugin);

  if (typeof plugin !== "function") {
    throw new Error(
      "phenomic: You provided an plugin with type is " +
        typeof plugin +
        ". But function is expected instead of " +
        String(plugin)
    );
  }

  // @todo send config here ?
  const pluginInstance = plugin();

  debug("plugin", pluginInstance.name);

  if (Array.isArray(pluginInstance)) {
    throw new Error(
      "Array of plugins should be specified in 'presets' section of your configuration"
    );
  }

  return pluginInstance;
};

const normalizeModule = (module: any) => {
  if (typeof module === "string") {
    // $FlowFixMe yeah yeah, I know what I am doing flow
    module = require(require.resolve(module));
  }

  // for es6 transpiled code
  if (module && typeof module.default === "function") {
    module = module.default;
  }

  debug("normalizeModule", module);

  return module;
};

function flattenPresets(config: PhenomicInputPlugins): PhenomicPlugins {
  debug("flattenPresets", config);
  const plugins = [
    ...(config.presets || [])
      .map(normalizeModule)
      .reduce((acc, preset) => [...acc, ...flattenPresets(preset())], []),
    ...(config.plugins || []).map(normalizeModule).map(normalizePlugin)
  ];
  debug("flattenPresets plugins", plugins);
  return plugins;
}

function flattenConfiguration(config: PhenomicInputConfig): PhenomicConfig {
  debug("flattenConfiguration", config);
  return {
    baseUrl: normalizeBaseUrl(config.baseUrl || defaultConfig.baseUrl),
    path: config.path || defaultConfig.path,
    content: config.content || defaultConfig.content,
    outdir: config.outdir || defaultConfig.outdir,
    port: config.port || defaultConfig.port,
    bundleName: config.bundleName || defaultConfig.bundleName,
    plugins: flattenPresets({
      plugins: config.plugins || [],
      presets: config.presets || []
    })
  };
}

export { flattenPresets };
export default flattenConfiguration;
