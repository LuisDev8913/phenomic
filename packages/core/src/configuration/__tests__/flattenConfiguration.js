// @flow

import { flattenPresets } from "../flattenConfiguration";

function a() {
  return { name: "a" };
}
function b() {
  return { name: "b" };
}
function c() {
  return { name: "c" };
}
function d() {
  return { name: "d" };
}

test("should flatten a list of plugins (array)", () => {
  expect(
    flattenPresets({
      plugins: [a, b]
    })
  ).toMatchSnapshot();
});

test("should flatten a list of plugins (object)", () => {
  expect(
    flattenPresets({
      plugins: { a, b }
    })
  ).toMatchSnapshot();
});

test("should flatten a list of plugins with options (array)", () => {
  expect(
    flattenPresets({
      plugins: [
        [(a: PhenomicPluginModule<PhenomicInputPluginOption>), { option: "-" }],
        b
      ]
    })
  ).toMatchSnapshot();
});

test("should flatten a list of plugins with options (object)", () => {
  expect(
    flattenPresets({
      plugins: {
        a: [
          (a: PhenomicPluginModule<PhenomicInputPluginOption>),
          { option: "-" }
        ],
        b
      }
    })
  ).toMatchSnapshot();
});

test("should flatten a preset (array)", () => {
  expect(
    flattenPresets({
      presets: [
        () => ({
          plugins: [a, b]
        })
      ]
    })
  ).toMatchSnapshot();
});

test("should flatten a preset (object)", () => {
  expect(
    flattenPresets({
      presets: [
        () => ({
          plugins: { a, b }
        })
      ]
    })
  ).toMatchSnapshot();
});

test("should flatten a preset and plugins", () => {
  expect(
    flattenPresets({
      presets: [
        () => ({
          plugins: [a]
        })
      ],
      plugins: [b, c]
    })
  ).toMatchSnapshot();
});

test("should flatten nested presets", () => {
  expect(
    flattenPresets({
      presets: [
        () => ({
          presets: [
            () => ({
              plugins: { a }
            })
          ],
          plugins: [b]
        })
      ],
      plugins: [c, d]
    })
  ).toMatchSnapshot();
});

test("should flatten presets with options (array)", () => {
  expect(
    flattenPresets({
      presets: [
        [
          () => ({
            plugins: { a, b, c }
          }),
          [["a", { option: "-" }]]
        ]
      ]
    })
  ).toMatchSnapshot();
});

test("should flatten presets with options (object)", () => {
  expect(
    flattenPresets({
      presets: [
        [
          () => ({
            plugins: { a, b, c }
          }),
          { a: { option: "-" } }
        ]
      ]
    })
  ).toMatchSnapshot();
});
