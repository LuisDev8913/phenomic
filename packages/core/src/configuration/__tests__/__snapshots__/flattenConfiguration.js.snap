// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flatten a list of plugins (array) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a list of plugins (object) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a list of plugins with options (array) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": Object {
      "option": "-",
    },
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a list of plugins with options (object) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": Object {
      "option": "-",
    },
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a preset (array) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a preset (object) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
]
`;

exports[`should flatten a preset and plugins 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "c",
    "options": undefined,
  },
]
`;

exports[`should flatten nested presets 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "c",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "d",
    "options": undefined,
  },
]
`;

exports[`should flatten presets with options (array) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": Object {
      "option": "-",
    },
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "c",
    "options": undefined,
  },
]
`;

exports[`should flatten presets with options (object) 1`] = `
Array [
  Object {
    "module": [Function],
    "name": "a",
    "options": Object {
      "option": "-",
    },
  },
  Object {
    "module": [Function],
    "name": "b",
    "options": undefined,
  },
  Object {
    "module": [Function],
    "name": "c",
    "options": undefined,
  },
]
`;
